### PR TITLE
Add Contour integral quadrature for exact GP sampling

### PR DESCRIPTION
**Pull request by jacobrgardner**
Implements contour integral quadrature as described **[here](https://arxiv.org/abs/2006.11267)** for sampling from exact GP posteriors. Any time `rsample` is called on a `MultivariateNormal` with a `LazyTensor` covariance matrix (or more specifically, when `LazyTensor.zero_mean_mvn_samples` is called), if it is done in a `with gpytorch.settings.ciq_samples():` context, the samples will be drawn using contour integral quadrature instead of Cholesky.

Support for whitened variational inference using CIQ will come in a follow up PR.

Setting is off by default because, like CG, it is not faster than Cholesky until N ~ 1000. Once the feature has been on master for a while, we can have it automatically switch to using CIQ in certain conditions. This will be slightly trickier, because Cholesky is faster than CIQ when drawing a very large number (e.g., O(N)) of samples.

### CODE REVIEW

File: gpytorch/lazy/lazy_tensor.py [hunk 1]
```diff
@@ -1585,15 +1621,33 @@ def zero_mean_mvn_samples(self, num_samples):
             :obj:`torch.tensor`:
                 Samples from MVN (num_samples x batch_size x num_dim) or (num_samples x num_dim)
         """
-        if self.size()[-2:] == torch.Size([1, 1]):
-            covar_root = self.evaluate().sqrt()
+        if settings.ciq_samples.on():
+            from ..utils.contour_integral_quad import contour_integral_quad
```
Comment (martinjankowiak) [inline gpytorch/lazy/lazy_tensor.py:1625]: why conditional importing?

File: gpytorch/lazy/lazy_tensor.py [hunk 2]
```diff
@@ -1585,15 +1621,33 @@ def zero_mean_mvn_samples(self, num_samples):
...
+
+            base_samples = torch.randn(
+                *self.batch_shape, self.size(-1), num_samples, dtype=self.dtype, device=self.device
+            )
+            base_samples = base_samples.permute(-1, *range(self.dim() - 1)).contiguous()
+            base_samples = base_samples.unsqueeze(-1)
+            solves, weights, _, _ = contour_integral_quad(
+                self.evaluate_kernel(),
```
Comment (martinjankowiak) [inline gpytorch/lazy/lazy_tensor.py:1633]: what's the plan for custom mvm support?

File: gpytorch/utils/cholesky.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 from .warnings import NumericalWarning
 
 
-def psd_safe_cholesky(A, upper=False, out=None, jitter=None):
+def psd_safe_cholesky(A, upper=False, out=None, jitter=None, max_tries=3):
```
Comment (martinjankowiak) [inline gpytorch/utils/cholesky.py:11]: add `max_tries` to docstring

File: gpytorch/utils/contour_integral_quad.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    :param torch.Tensor rhs: Right hand side tensor :math:`\mathbf b`
+    :param bool inverse: (default False) whether to compute :math:`\mathbf K^{1/2} \mathbf b` (if False)
+        or `\mathbf K^{-1/2} \mathbf b` (if True)
+    :param int max_lanczos_iter: (default 10) Number of Lanczos iterations to run (to estimate eigenvalues)
+    :param int num_contour_quadrature: (default 15) How many quadrature samples to use for approximation
+    :rtype: torch.Tensor
+    :return: Approximation to :math:`\mathbf K^{1/2} \mathbf b` or :math:`\mathbf K^{-1/2} \mathbf b`.
+    """
+    import numpy as np
+    from scipy.special import ellipj, ellipk
```
Comment (martinjankowiak) [inline gpytorch/utils/contour_integral_quad.py:39]: i take it this is already a gpytorch requirement?

File: gpytorch/utils/contour_integral_quad.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        # Compute an approximate condition number
+        # We'll do this with Lanczos
+        try:
+            approx_eigs = lanczos_mat.symeig()[0]
+            if approx_eigs.min() <= 0:
+                raise RuntimeError
+        except RuntimeError:
+            approx_eigs = lazy_tensor.diag()
+
+        max_eig = approx_eigs.max(dim=-1)[0]
```
Comment (martinjankowiak) [inline gpytorch/utils/contour_integral_quad.py:87]: should we make these more conserative?

File: gpytorch/utils/contour_integral_quad.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    max_lanczos_iter=20,
+    num_contour_quadrature=7,
+    shift_offset=0,
+):
+    r"""
+    Performs :math:`\mathbf K^{1/2} \mathbf b` or `\mathbf K^{-1/2} \mathbf b`
+    using contour integral quadrature.
+
+    .. note::
+        Right now this only works for non-batch matrices
```
Comment (martinjankowiak) [inline gpytorch/utils/contour_integral_quad.py:27]: will this fail gracefully?

File: gpytorch/utils/minres.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        uses the constant stored in :obj:`gpytorch.settings.max_cg_iterations`.
+    :rtype: torch.Tensor
+    :return: The solves :math:`\mathbf x`. The shape will correspond to the size of `rhs` and `shifts`.
+    """
+    # Default values
+    if torch.is_tensor(matmul_closure):
+        matmul_closure = matmul_closure.matmul
+    mm_ = matmul_closure
+    if preconditioner is None:
+        preconditioner = lambda x: x.clone()
```
Comment (martinjankowiak) [inline gpytorch/utils/minres.py:30]: why clone?

File: gpytorch/utils/minres.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    # Epsilon (to prevent nans)
+    eps = torch.tensor(eps, dtype=rhs.dtype, device=rhs.device)
+
+    # Create space for matmul product, solution
+    prod = mm_(rhs)
+    if value is not None:
+        prod.mul_(value)
+
+    # Reisze shifts
```
Comment (martinjankowiak) [inline gpytorch/utils/minres.py:59]: typo

File: gpytorch/utils/minres.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    shifts = _pad_with_singletons(shifts, 0, prod.dim() - shifts.dim() + 1)
+    solution = torch.zeros(shifts.shape[:1] + prod.shape, dtype=rhs.dtype, device=rhs.device)
+
+    # Variasbles for Lanczos terms
```
Comment (martinjankowiak) [inline gpytorch/utils/minres.py:63]: typo

File: gpytorch/lazy/lazy_tensor.py (hunk 2 above)
Comment (jacobrgardner) [reply gpytorch/lazy/lazy_tensor.py:1633]: This just evaluates any outstanding kernel calls, it doesn't necessarily explicitly materialize an `n x n` matrix (i.e., `evaluate_kernel` vs `evaluate`).

File: gpytorch/utils/contour_integral_quad.py (hunk 4 above)
Comment (jacobrgardner) [reply gpytorch/utils/contour_integral_quad.py:39]: Actually I looked and we explicitly don't require you to have numpy / scipy installed anywhere and have by convention imported them directly in functions/features where they are used (the same way we handle pyro). I ended up moving the imports down here to follow this convention

File: gpytorch/utils/contour_integral_quad.py (hunk 5 above)
Comment (jacobrgardner) [reply gpytorch/utils/contour_integral_quad.py:87]: I don't know how we would do this in a principled way. For many matrices, our condition number estimate will be accurate to machine precision.

File: gpytorch/utils/minres.py (hunk 7 above)
Comment (jacobrgardner) [reply gpytorch/utils/minres.py:30]: We later modify the output of the preconditioner in place:
https://github.com/cornellius-gp/gpytorch/blob/fac03c2622e72288e2edf5d5b3edce7cbc7b3846/gpytorch/utils/minres.py#L75

File: gpytorch/utils/contour_integral_quad.py (hunk 6 above)
Comment (jacobrgardner) [reply gpytorch/utils/contour_integral_quad.py:27]: Oops, that note actually isn't even true anymore

File: gpytorch/lazy/lazy_tensor.py (hunk 2 above)
Comment (martinjankowiak) [reply gpytorch/lazy/lazy_tensor.py:1633]: ah gotcha thanks

File: gpytorch/utils/contour_integral_quad.py (hunk 5 above)
Comment (martinjankowiak) [reply gpytorch/utils/contour_integral_quad.py:84]: why not just do a factor of 2 fudge or something?

File: gpytorch/settings.py [hunk 10]
```diff
@@ -484,6 +511,14 @@ class prior_mode(_feature_flag):
...
 
 
+class single_precision_cholesky(_feature_flag):
+    """
+    Use single precision when computing the Cholesky decomposition (and solves)
+    """
+
+    _state = False
+
+
```
Comment (jacobrgardner) [inline gpytorch/settings.py:521]:
```suggestion
```
This setting is currently unused on any branch as far as I can tell

File: gpytorch/utils/contour_integral_quad.py (hunk 5 above)
Comment (jacobrgardner) [reply gpytorch/utils/contour_integral_quad.py:84]: Just multiplying the max eigenvalue by 2 almost surely gets us further from the true max eigenvalue than the estimate here, and I'm not clear that intentionally overestimating the condition number by a factor of sqrt(2) buys us anything in terms of CIQ convergence? 

These estimates are going to be significantly more accurate than, e.g., 20 iterations of power iteration. They should be pretty darn near exact.

File: gpytorch/utils/contour_integral_quad.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+
+def contour_integral_quad(
+    lazy_tensor,
+    rhs,
+    inverse=False,
+    weights=None,
+    shifts=None,
+    max_lanczos_iter=20,
+    num_contour_quadrature=7,
```
Comment (dme65) [inline gpytorch/utils/contour_integral_quad.py:19]: Default should be 15?

**Pull request merged**